### PR TITLE
enable sub-solutions

### PR DESCRIPTION
**Pull request by malb**
- add sub-solution support
- make API consistent between sub-solutions and full solutions/fplll

### CODE REVIEW

File: src/fpylll/fplll/enumeration.pyx [hunk 1]
```diff
@@ -433,11 +435,47 @@ cdef class Enumeration:
...
+        >>> from fpylll import *
+        >>> set_random_seed(1337)
+        >>> A = IntegerMatrix.random(80, "qary", bits=30, k=40)
+        >>> _ = LLL.reduction(A)
+        >>> M = GSO.Mat(A)
+        >>> _ = M.update_gso()
+        >>> pruning = Pruning.run(M.get_r(0, 0), 2**40, M.r()[:30], 0.2)
+        >>> enum = Enumeration(M, strategy=EvaluatorStrategy.BEST_N_SOLUTIONS, sub_solutions=True)
+        >>> _ = enum.enumerate(0, 30, 0.999*M.get_r(0, 0), 0, pruning=pruning.coefficients)
+        >>> [int(a) for a,b in enum.sub_solutions[:5]]
```
Comment (malb) [inline src/fpylll/fplll/enumeration.pyx]: Tests are here, could check for more mathematical properties, though

**Comment by malb** [comment]
We should wait with merging until the possible unallocated memory issue is resolved. `set_random_seed` doesn't allow to reproduce it?

File: src/fpylll/fplll/enumeration.pyx (hunk 1 above)
Comment (lducas) [inline src/fpylll/fplll/enumeration.pyx]: Ta.

**Comment by lducas** [comment]
Its actually not such a huge bug. It is just a rather awkward behaviour. When nothing is found two things can happen:
1/ (a, b) = (0.0, (0.0, ... 0.0))
2/ (a, b) = (random, ())

So I guess initialized memory is involved in case 2, yet nothing too deep I'd guess. Not sure if this should be fixed in fplll or fpylll.

**Comment by lducas** [comment]
For any practical purpose, the attached pickle files contains a triple 
```(B, radius, pruning)'''
where case 2/ happened (pruned enumeration on the whole basis B).
[buugy_inst.pkl.zip](https://github.com/fplll/fpylll/files/1203013/buugy_inst.pkl.zip)

**Pull request merged**